# Lab 5.04 - Two-sample t-test

In [37]:
# Package imports
import numpy as np                                  # "Scientific computing"
import scipy.stats as stats                         # Statistical tests

import pandas as pd                                 # Dataframe
import matplotlib.pyplot as plt                     # Basic visualisation
from statsmodels.graphics.mosaicplot import mosaic  # Mosaic plot
import seaborn as sns                               # Advanced dataviz

## Exercise 4 - Android Persistence libraries performance comparison

We analyzed the results of performance measurements for Android persistence libraries (Akin, 2016). Experiments were performed for different combinations of *DataSize* (Small, Medium, Large) and *PersistenceType* (GreenDAO, Realm, SharedPreferences, SQLite). For each data size, we were able to determine which persistence type yielded the best results.

Now we will verify if the best persistence type at first glance is also *significantly* better than the competition.

Specifically: Using a two-sample test for each data size, verify that the mean of the best persistence type is significantly lower than the mean of the second best and the worst scoring type.

Can we maintain the conclusion that for a given data size, one persistence type is best, i.e. is significantly better than any other persistence type?

In [54]:
df = pd.read_csv('../data/android_persistence_cpu.csv', delimiter=';')
df.head()

Time    PersistenceType DataSize
0  1.81  Sharedpreferences    Small
1  1.35  Sharedpreferences    Small
2  1.84  Sharedpreferences    Small
3  1.54  Sharedpreferences    Small
4  1.81  Sharedpreferences    Small

In [56]:
help = df.groupby(['DataSize', 'PersistenceType'])['Time'].describe().reset_index()
help.head()

DataSize PersistenceType  count       mean       std    min      25%  \
0    Large        GreenDAO   30.0  12.110333  0.868461  10.38  11.4775   
1    Large           Realm   30.0  10.651667  1.405990   7.68   9.6700   
2    Large         SQLLite   30.0  11.515000  0.559142  10.58  11.1175   
3   Medium        GreenDAO   30.0   7.454000  2.007358   4.18   6.0700   
4   Medium           Realm   30.0   5.818000  1.330951   3.79   4.8925   

      50%      75%    max  
0  12.045  12.7000  13.56  
1  10.660  11.8275  13.36  
2  11.425  11.7800  12.96  
3   7.425   8.7150  11.94  
4   5.690   6.4125  10.34

In [58]:
stats.ttest_ind(a=df[(df.DataSize =='Small') & (df.PersistenceType =='Realm' )][ 'Time'], b=df[ (df.DataSize == 'Small') & (df.PersistenceType == 'Sharedpreferences' )]['Time'], alternative='less', equal_var=False)

TtestResult(statistic=-0.9624716662718156, pvalue=0.16992370571901444, df=57.43660193307136)

### Answers

The table below provides an overview of the best and second best persistence type for each data size (based on the sample mean).

| Data Size | Best  | 2nd Best          | p-value   |
| :-------- | :---- | :---------------- | :-------- |
| Small     | Realm | SharedPreferences | 0.1699    |
| Medium    | Realm | GreenDAO          | 0.0002506 |
| Large     | Realm | SQLite            | 0.0017    |

The conclusion of Akin (2016), which states that Realm is the most efficient persistence type, still holds, but for the small data sets the difference is not significant.

Note that we have not explicitly selected a specific significance level in advance. However, for $\alpha$ = 0.1, 0.05 or even 0.01, the same conclusion can be drawn.